File: src/main/strategies/launch-strategy.ts
import { spawn } from "node:child_process";
import { PathStrategyFactory, ProtonPathStrategy } from "./path-strategy";
import { detectSteamAppId } from "../utils/game-detector";

export interface LaunchStrategy {
  launch(
    executablePath: string,
    modPaths: string[],
    gameDir: string,
    extraArgs?: string[]
  ): void;
  getLaunchOptions(modPaths: string[], extraArgs?: string[]): string;
}

class DirectLaunchStrategy implements LaunchStrategy {
  launch(
    executablePath: string,
    modPaths: string[],
    gameDir: string,
    extraArgs: string[] = []
  ): void {
    const args = [executablePath, ...extraArgs];
    if (modPaths.length > 0) {
      args.push("-modpaths", ...modPaths);
    }
    spawn(args[0], args.slice(1), { cwd: gameDir, detached: true, stdio: "ignore" });
  }

  getLaunchOptions(modPaths: string[], extraArgs: string[] = []): string {
    const parts = [...extraArgs];
    if (modPaths.length > 0) {
      parts.push("-modpaths", ...modPaths.map((p) => `"${p}"`));
    }
    return parts.join(" ");
  }
}

class ProtonLaunchStrategy implements LaunchStrategy {
  private readonly appId: string;

  constructor(private readonly gameDir: string) {
    this.appId = detectSteamAppId(gameDir);
  }

  launch(
    executablePath: string,
    modPaths: string[],
    gameDir: string,
    extraArgs: string[] = []
  ): void {
    if (process.platform === "linux") {
      this.launchViaSteam(modPaths, extraArgs);
    } else {
      new DirectLaunchStrategy().launch(
        executablePath,
        modPaths,
        gameDir,
        extraArgs
      );
    }
  }

  private launchViaSteam(modPaths: string[], extraArgs: string[]): void {
    if (!this.appId) {
      throw new Error(
        "Could not detect Steam App ID. Please ensure the game is installed through Steam."
      );
    }

    const launchOptions = [...extraArgs];
    if (modPaths.length > 0) {
      launchOptions.push("-modpaths", ...modPaths);
    }

    try {
      const cmd = ["steam", "-applaunch", this.appId, ...launchOptions];
      spawn(cmd[0], cmd.slice(1), { detached: true, stdio: "ignore" });
    } catch {
      spawn("xdg-open", [`steam://rungameid/${this.appId}`], {
        detached: true,
        stdio: "ignore",
      });
    }
  }

  getLaunchOptions(modPaths: string[], extraArgs: string[] = []): string {
    return new DirectLaunchStrategy().getLaunchOptions(modPaths, extraArgs);
  }
}

export class LaunchStrategyFactory {
  static create(gameDir: string): LaunchStrategy {
    if (PathStrategyFactory.isProton(gameDir)) {
      return new ProtonLaunchStrategy(gameDir);
    }
    return new DirectLaunchStrategy();
  }
}

export { ProtonPathStrategy };

File: .mewtator/app/core/strategies/launch_strategy.py
from abc import ABC, abstractmethod
from typing import List
import os
import sys
import subprocess


class LaunchStrategy(ABC):
    @abstractmethod
    def launch(self, executable_path: str, mod_paths: List[str], game_dir: str, extra_args: List[str] = None):
        pass
    
    @abstractmethod
    def get_launch_options(self, mod_paths: List[str], extra_args: List[str] = None) -> str:
        pass


class DirectLaunchStrategy(LaunchStrategy):
    def launch(self, executable_path: str, mod_paths: List[str], game_dir: str, extra_args: List[str] = None):
        args = [executable_path]
        
        if extra_args:
            args.extend(extra_args)
        
        if mod_paths:
            args.append("-modpaths")
            args.extend(mod_paths)
        
        subprocess.Popen(args, cwd=game_dir)
    
    def get_launch_options(self, mod_paths: List[str], extra_args: List[str] = None) -> str:
        parts = []
        
        if extra_args:
            parts.extend(extra_args)
        
        if mod_paths:
            parts.append("-modpaths")
            parts.extend(f'"{p}"' for p in mod_paths)
        
        return " ".join(parts)


class ProtonLaunchStrategy(LaunchStrategy):
    def __init__(self, path_converter, game_dir: str):
        self.path_converter = path_converter
        self.game_dir = game_dir
        self.app_id = self._detect_steam_app_id()
    
    def _detect_steam_app_id(self) -> str:
        """Detect Steam App ID for the game."""
        from app.utils.game_detector import detect_steam_app_id
        return detect_steam_app_id(self.game_dir)
    
    def launch(self, executable_path: str, mod_paths: List[str], game_dir: str, extra_args: List[str] = None):
        if sys.platform.startswith('linux'):
            self._launch_via_steam(executable_path, mod_paths, game_dir, extra_args)
        else:
            self._launch_direct(executable_path, mod_paths, game_dir, extra_args)
    
    def _launch_via_steam(self, executable_path: str, mod_paths: List[str], game_dir: str, extra_args: List[str] = None):
        """Launch the game through Steam on Linux."""
        if not self.app_id:
            raise RuntimeError(
                "Could not detect Steam App ID. Please ensure the game is installed through Steam.\n\n"
                "To launch with mods on Linux, you need to:\n"
                "1. Use 'Copy Launch Options' from the File menu\n"
                "2. Paste them in Steam: Right-click Mewgenics → Properties → Launch Options\n"
                ""
                "3. Launch the game from Steam"
            )
        
        launch_options = []
        
        if extra_args:
            launch_options.extend(extra_args)
        
        if mod_paths:
            launch_options.append("-modpaths")
            launch_options.extend(mod_paths)
        
        try:
            cmd = ['steam', '-applaunch', self.app_id]
            if launch_options:
                cmd.extend(launch_options)
            subprocess.Popen(cmd, start_new_session=True)
        except FileNotFoundError:
            try:
                steam_uri = f"steam://rungameid/{self.app_id}"
                subprocess.Popen(['xdg-open', steam_uri], start_new_session=True)
            except FileNotFoundError:
                raise RuntimeError(
                    "Could not launch Steam. Please ensure Steam is installed and running.\n\n"
                    f"You can manually launch with: steam -applaunch {self.app_id}"
                )
    
    def _launch_direct(self, executable_path: str, mod_paths: List[str], game_dir: str, extra_args: List[str] = None):
        """Direct launch (fallback for non-Linux platforms)."""
        args = [executable_path]
        
        if extra_args:
            args.extend(extra_args)
        
        if mod_paths:
            args.append("-modpaths")
            args.extend(mod_paths)
        
        subprocess.Popen(args, cwd=game_dir)
    
    def get_launch_options(self, mod_paths: List[str], extra_args: List[str] = None) -> str:
        parts = []
        
        if extra_args:
            parts.extend(extra_args)
        
        if mod_paths:
            parts.append("-modpaths")
            parts.extend(f'"{p}"' for p in mod_paths)
        
        return " ".join(parts)


class LaunchStrategyFactory:
    @staticmethod
    def create(game_dir: str) -> LaunchStrategy:
        from app.core.strategies.path_strategy import PathStrategyFactory, ProtonPathStrategy
        
        path_strategy = PathStrategyFactory.create(game_dir)
        
        if isinstance(path_strategy, ProtonPathStrategy):
            return ProtonLaunchStrategy(ProtonPathStrategy._convert_to_proton_path, game_dir)
        
        return DirectLaunchStrategy()
